caesar/caesar.6
.\"	$NetBSD: caesar.6,v 1.4 1995/03/21 15:08:20 cgd Exp $
.\"
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)caesar.6	8.2 (Berkeley) 11/16/93
.\"
.Dd November 16, 1993
.Dt CAESAR 6
.Os
.Sh NAME
.Nm caesar
.Nd decrypt caesar cyphers
.br
.Nm rot13
.Nd decrypt rot13ed Usenet posts
.Sh SYNOPSIS
.Nm caesar
.Op Ar rotation
.br
.Nm rot13
.Sh DESCRIPTION
The
.Nm caesar
utility attempts to decrypt caesar cyphers using English letter frequency
statistics.
.Nm Caesar
reads from the standard input and writes to the standard output.
.Pp
The optional numerical argument
.Ar rotation
may be used to specify a specific rotation value.
.Pp
The frequency (from most common to least) of English letters is as follows:
.Pp
.Bd -filled -offset indent
ETAONRISHDLFCMUGPYWBVKXJQZ
.Ed
.Pp
Their frequencies as a percentage are as follows:
.Pp
.Bd -filled -offset indent
E(13), T(10.5), A(8.1), O(7.9), N(7.1), R(6.8), I(6.3), S(6.1), H(5.2),
D(3.8), L(3.4), F(2.9), C(2.7), M(2.5), U(2.4), G(2),
P(1.9), Y(1.9),
W(1.5), B(1.4), V(.9), K(.4), X(.15), J(.13), Q(.11), Z(.07).
.Ed
.Pp
Rotated postings to
.Tn USENET
and some of the databases used by the
.Xr fortune 6
program are rotated by 13 characters.


NEWS
NEWS for bsd-games and bsd-games-non-free
=========================================

This file contains a summary of changes in each version of bsd-games
and bsd-games-non-free after 1.3.  For fuller details, see ChangeLog
and ChangeLog.0.

As well as reading this summary, it is a good idea to reread the
INSTALL file whenever installing a new version, to check for any
changes in installation instructions.  Packagers should also reread
the PACKAGING file each version (or do a diff between the old and new
versions of the package and read that).

Version 2.8
===========

* Robots scorefile format has changed - remove or rename your old
score file when upgrading to this version.

* Various minor changes and bug-fixes (some from OpenBSD).

* New `game': wtf.

* Updated from NetBSD CVS of yyyy-mm-dd; many more changes merged back
into NetBSD.

* All constant data files are now architecture independent (installed
under /usr/share by default).

* bsd-games is no longer tested by me with libc older than glibc
2.1.2; fixes for compilation with older libc will still be accepted.

* Details of the copyrights and licence terms applicable to these
games, as best as I can determine them, now appear in COPYING.

* bsd-games now has a testsuite (for programs that can be tested
non-interactively only).

Version 2.7
===========

* Bug fixes.

* Updated from NetBSD-current of 1999-03-26; more changes merged back
into NetBSD.

* Other minor changes.

Version 2.6
===========

* Buffer overrun fixes in sail (not thought exploitable).

* Fixes possible score file corruption if fds 0, 1, 2 aren't open at
game startup.

* Updated from NetBSD-current of 1998-12-07 (morse now has a decoding
option).

* Now builds cleanly with -Wwrite-strings.

* Other minor changes.

* Some users of this package may be interested in my `nppt' package,
which provides an alternative implementation of ppt(6) with additional
features such as Encapsulated PostScript output.  (This package
installs the binary as `nppt', so can be installed simultaneously with
this one.  Look for nppt-0.1.tar.gz (or a later version) on
metalab.unc.edu (formerly sunsite) in /pub/Linux/Incoming/ or under
/pub/Linux/games/.)

Version 2.5
===========

* New build system, using a single project Makefile (autogenerated)
that includes makefile fragments from subdirectories, and uses GNU
Make's automatic dependency generation to generate dependencies
automatically, rather than recursive make.  See Peter Miller's paper
`Recursive Make Considered Harmful'
<http://www.canb.auug.org.au/~millerp/rmch/recu-make-cons-harm.html>
for reasons why this is better than recursive make.  This should make
no difference to normal users, but if you modify the sources note that
you will need to rebuild from the top directory, and you may find that
the dependencies are now accurate whereas formerly they were
incomplete.  The automatic dependency generation causes the build to
start with about 300 `no such file or directory' warnings, which
should be ignored: these refer to the dependency files which GNU Make
will then automatically generate.

* Should work with glibc 2.1 (at least, version 2.0.96 on x86, which
is what I've tested).  See the ChangeLog for an explanation of the
problem previous releases had with glibc 2.1.

* Boggle is now in the main bsd-games package - Paul Janzen checked
with the original author, who confirmed that the BSD licence
superseded the old non-free one.  bsd-games-non-free now only contains
rogue.

* Other minor changes.

Version 2.4
===========

* Various minor bug fixes and improvements (some from OpenBSD, thanks
to Paul Janzen <pjanzen@foatdi.harvard.edu>).

* Sail directory can now be non-world-accessible (mode 2770 root.games
or similar).  This may improve security, since it means users cannot
tamper with sync files their sail process creates.

* Rogue buffer overrun fix.  An audit is still needed.

* dm now ignores the environment variable TZ when deciding at what
times games can be played.  (This only affects you if you are using dm
to control at what times games can be played, and not then if you are
using libc 5.4.45 or 5.4.46 (which completely ignore TZ in setgid
programs rather than checking the validity of any filename given as
glibc does).)

* All games updated from NetBSD-current of 1998-08-30 (minor changes;
some more of my changes have been merged back into NetBSD).

Version 2.3
===========

* A possible temporary file security bug in sail fixed.

* Countmail is back, with a licence (apparently the standard NetBSD
one), since it has one in NetBSD-current of 1998-08-19.

* All the other games updated from NetBSD-current of 1998-08-19 (minor
changes).

* Other minor changes.

* Version 2.2 was never properly released, since the upload to sunsite
was truncated and the sail hole was found before the truncated file
had been removed or the release put in place on tsx-11.

Version 2.2
===========

* Another bug-fix release.

* This version includes security fixes from OpenBSD.

* This release is believed year 2000 compliant.  Please see the year
2000 statement in the file YEAR2000 for details and disclaimer.

* Paranoia has been removed from bsd-games-non-free, since it has no
clear licence at all, and being derived from a magazine article it is
not clear it was ever legally distributable.  Anyone wanting to
resurrect it in a separate package would need to investigate the
copyright on the magazine article as well as getting a proper licence
from the author.

* Countmail has been removed from bsd-games because of the lack of a
licence.

* All other games updated from NetBSD-current of 1998-07-27 (very
minor changes).

* Default paths now follow FHS 2.0 - enter the paths you want when
running configure if you are using the FSSTND 1.2.

* You can now easily specify games you don't want built when running
configure.

* All games that can invoke pagers handle PAGER in the way specified
by POSIX.2 for standard programs that invoke pagers.


caesar/caesar.c
/*	$NetBSD: caesar.c,v 1.12 2003/08/07 09:37:07 agc Exp $	*/

/*
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Rick Adams.
 *
 * Authors:
 *	Stan King, John Eldridge, based on algorithm suggested by
 *	Bob Morris
 * 29-Sep-82
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1989, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)caesar.c	8.1 (Berkeley) 5/31/93";
#else
__RCSID("$NetBSD: caesar.c,v 1.12 2003/08/07 09:37:07 agc Exp $");
#endif
#endif /* not lint */

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>

#define	LINELENGTH	2048
#define	ROTATE(ch, perm) \
	isupper(ch) ? ('A' + (ch - 'A' + perm) % 26) : \
	    islower(ch) ? ('a' + (ch - 'a' + perm) % 26) : ch

/*
 * letter frequencies (taken from some unix(tm) documentation)
 * (unix is a trademark of Bell Laboratories)
 */
double stdf[26] = {
	7.97, 1.35, 3.61, 4.78, 12.37, 2.01, 1.46, 4.49, 6.39, 0.04,
	0.42, 3.81, 2.69, 5.92,  6.96, 2.91, 0.08, 6.63, 8.77, 9.68,
	2.62, 0.81, 1.88, 0.23,  2.07, 0.06,
};


int	main __P((int, char *[]));
void	printit __P((const char *)) __attribute__((__noreturn__));

int
main(argc, argv)
	int argc;
	char **argv;
{
	int ch, dot, i, nread, winnerdot;
	char *inbuf;
	int obs[26], try, winner;

	/* revoke setgid privileges */
	setregid(getgid(), getgid());

	winnerdot = 0;
	if (argc > 1)
		printit(argv[1]);

	if (!(inbuf = malloc(LINELENGTH)))
		err(1, NULL);

	/* adjust frequency table to weight low probs REAL low */
	for (i = 0; i < 26; ++i)
		stdf[i] = log(stdf[i]) + log(26.0 / 100.0);

	/* zero out observation table */
	memset(obs, 0, 26 * sizeof(int));

	if ((nread = read(STDIN_FILENO, inbuf, LINELENGTH)) < 0)
		err(1, "reading from stdin");
	for (i = nread; i--;) {
		ch = inbuf[i];
		if (islower(ch))
			++obs[ch - 'a'];
		else if (isupper(ch))
			++obs[ch - 'A'];
	}

	/*
	 * now "dot" the freqs with the observed letter freqs
	 * and keep track of best fit
	 */
	for (try = winner = 0; try < 26; ++try) { /* += 13) { */
		dot = 0;
		for (i = 0; i < 26; i++)
			dot += obs[i] * stdf[(i + try) % 26];
		/* initialize winning score */
		if (try == 0)
			winnerdot = dot;
		if (dot > winnerdot) {
			/* got a new winner! */
			winner = try;
			winnerdot = dot;
		}
	}

	for (;;) {
		for (i = 0; i < nread; ++i) {
			ch = inbuf[i];
			putchar(ROTATE(ch, winner));
		}
		if (nread < LINELENGTH)
			break;
		if ((nread = read(STDIN_FILENO, inbuf, LINELENGTH)) < 0)
			err(1, "reading from stdin");
	}
	exit(0);
}

void
printit(arg)
	const char *arg;
{
	int ch, rot;

	if ((rot = atoi(arg)) < 0)
		errx(1, "bad rotation value.");
	while ((ch = getchar()) != EOF)
		putchar(ROTATE(ch, rot));
	exit(0);
}
